### DISCUSSION TITLE
Summation of Element

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#19274] [Q&A General] [answered]
Hello, I have a question about built-in Kernel or Postprocessor in MOOSE.

I want to sum values stored (assigned) in all Elements.

Is there any proper Kernel or Postprocessor (or AuxKernel? ) for this?

I tried to find an appropriate one through Syntax, however I didn't know which one is right for me.

Thanks!

Sincerely,
Hokon Kim

**Comment by GiudGiud** [comment, marked as answer]
Hello

So you want a sum with no quadrature? Like not an integration? I m afraid we don't have that already in the framework. 

A sum would be a single value, so it would be a postprocessor. 

If all your elements are the same shape you can use an ElementIntegralVariablePostprocessor since the integral and the sum are the same
https://mooseframework.inl.gov/source/postprocessors/ElementIntegralVariablePostprocessor.html
If all your elements have the same number of nodes, you can use a NodalSum, as this does not do an integration, just sums the values on each node, which for the right type of variable will just be equal to the element value. Then you can divide the result by the number of nodes per element. 

We would absolutely support it if you wanted to add an ElementSum post processor to the framework. 

Guillaume

**Reply by hokonkim** [reply]
![image](https://user-images.githubusercontent.com/46009457/139878629-26fda84e-a06b-4613-9f3d-fec9eef0e43c.png)

For example, above image is the 3D microstructure that had Elements(colored as Black points).
(I do not know how many elements or nodes exists, how could I know this?)

I just want to sum values(colored as red) on black points at each (time) step.
-> I didn't give all numbers on black points, but values are assigned on all Elements

Is it possible to compute this summation?

**Reply by GiudGiud** [reply]
number of elements and nodes is output at code startup / you can know it from the mesh using a viewer

no not as a summation unless you are in one of the special cases I mentioned. 
You d need to write a new post processor if not

**Reply by hokonkim** [reply]
Here is the mesh information that I used recently.

    Mesh: 
      Parallel Type:           replicated
      Mesh Dimension:          3
      Spatial Dimension:       3
      Nodes:                   
        Total:                 1021076
        Local:                 9296
      Elems:                   
        Total:                 5708148
        Local:                 47605
      Num Subdomains:          4
      Num Partitions:          120
      Partitioner:             metis

I think I need to write a new post-processor for what I want.

What am I supposed to do?

Thanks for your help.

**Reply by GiudGiud** [reply]
I would start from ElementIntegralVariablePP (so copy it to ElementSum, then modify it)
and instead of summing on the quadrature with weights, I would only sum if qp = 0 with a weight of 1

**Reply by hokonkim** [reply]
So, you mean I need to write a new Post-Processor like I made Custom kernels or interfacekernels for my application.

Also, it would be good to start from copying ElementsIntegralVariablePP as ElementSum(Custom Post processor for my application).

Am I right on track?

**Reply by GiudGiud** [reply]
yes on both counts

**Reply by GiudGiud** [reply]
once you've created it I d be down to have it in the framework, not just in your app though

**Reply by hokonkim** [reply]
I copied and pasted the ElementIntegralVariablePP.C and ElementIntegralVariablePP.h from MOOSE framework PP folder.

and replace ElementIntegralVariablePP with ElementSumPP.

But, still I have no idea about implementing this part "instead of summing on the quadrature with weights, I would only sum if qp = 0 with a weight of 1" below.

    Real
    ElementSumPostprocessor::computeQpIntegral()
    {
      return _u[_qp];
    }

[ElementSumPostprocessor.zip](https://github.com/idaholab/moose/files/7463120/ElementSumPostprocessor.zip)

**Reply by GiudGiud** [reply]
the weight isn't in computeQpIntegral it s in computeIntegral I think

In computeQpIntegral, for selecting a Qp, you can just have:
if _qp == 0
  return _u[_qp];
else
  return 0

Note that I am assuming that the variable is constant over ELEMENTS and that you want one value per element.
From your drawing it looks like you want a NodalSum ??

**Reply by hokonkim** [reply]
Okay, I think I need to give you more detail.

what I want is to sum values in Elements (specifically, "Auxiliary variable", not "Variable").

Depending on summation, I would control the "Material property" gradually.

For NodalSum that you commented previously, my mesh does not have equal numbers for Node and Element.
(_If all your elements have the same number of nodes, you can use a NodalSum_)

So, I considered writing the new PP like you said to me.



    Real
    ElementSumPostprocessor::computeQpIntegral()
    {
      if _qp == 0{
        return _u[_qp];
      }
      else{
        return 0;
      }
    }

**Reply by GiudGiud** [reply]
ok so the computeIntegral routine, which lives in ElementIntegralPostprocessor, is still adding the volume weighting
see here 
https://github.com/idaholab/moose/blob/next/framework/src/postprocessors/ElementIntegralPostprocessor.C

you ll need to override it


For a better design, I think you will want to start from ElementIntegralPP not ElementIntegralVariablePP. as you want

**Reply by hokonkim** [reply]
I started from ElementIntegralPP like you said.

In "ElementIntegralPP.C", I overwrote like below

    Real
    ElementSumPostprocessor::computeIntegral()
    {
      // Real sum = 0;
      //
      // for (_qp = 0; _qp < _qrule->n_points(); _qp++)
      //   sum += _JxW[_qp] * _coord[_qp] * computeQpIntegral();
      // return sum;

      if _qp == 0{
        return _u[_qp];
      }
      else{
        return 0;
      }
    }

[ElementSumPostprocessor.zip](https://github.com/idaholab/moose/files/7463336/ElementSumPostprocessor.zip)

Did I do right?

**Reply by GiudGiud** [reply]
probably need (_qp == 0) but otherwise if it does well on a test it seems right

**Reply by GiudGiud** [reply]
well actually _qp was defined by that loop.
So just return _u[0] no need to look a the _qp here

**Reply by hokonkim** [reply]
This is the latest version for ElementSumPP.

    Real
    ElementSumPostprocessor::computeIntegral()
    {
      if (_qp == 0){
        return _u[0];
      }
      else{
        return 0;
      }
    }

[ElementSumPostprocessor.zip](https://github.com/idaholab/moose/files/7463787/ElementSumPostprocessor.zip)

I will test this PP as soon as possible and need to make another AuxKernel, but I have no issue for now.

So, like custom kernels, re-compile is needed to use this PP on my application, right?

**Reply by GiudGiud** [reply]
yes

**Reply by hokonkim** [reply]
I updated MOOSE and I tried to recompile my application with new postprocessor I made.

But I got an error like below
```
    In file included from /.nfs/home/6/kimho/projects/ermine/build/unity_src/postprocessors_Unity.C:2:
    /.nfs/home/6/kimho/projects/ermine/src/postprocessors/ElementSumPostprocessor.C: In member function \u2018virtual libMesh::Real 
    ElementSumPostprocessor::computeIntegral()\u2019:
    /.nfs/home/6/kimho/projects/ermine/src/postprocessors/ElementSumPostprocessor.C:58:12: error: \u2018_u\u2019 was not declared in 
    this scope; did you mean \u2018_t\u2019?
       58 |     return _u[0];
          |            ^~
          |            _t
```
What should I do to fix this issue?

**Reply by GiudGiud** [reply]
I dont think updating moose created that issue. It was there before.

The issue is that you are inheriting from ElementPostprocessor, which does not define the variable. Even ElementIntegralPostprocessor does not. The first one in the chain for a regular integral to add the variable as a member of the class is ElementIntegralVariablePostprocessor. 

so either:
- inherit from ElementIntegralVariablePostprocessor and make sure to override the integration code (otherwise the weights will come in) from ElementIntegralpostprocessor
OR
- look at the code for ElementIntegralVariablePostprocessor and add the relevant parts for you. A variable interface is used. It should be a few copy pastes to get that code in and define _u

**Reply by hokonkim** [reply]
I tried to do the second way...

For header file (ElementSumPostprocessor.h),

    #pragma once
    
    #include "ElementPostprocessor.h"
    #include "MooseVariableInterface.h"

    // Forward Declarations
    class ElementSumPostprocessor;

    template <>
    InputParameters validParams<ElementSumPostprocessor>();

    /**
     * This postprocessor computes a summation of values assigned in Elements.
     *
     * Note that specializations of this summation are possible by deriving from this
     * class and overriding computeQpIntegral().
     */
    class ElementSumPostprocessor : public ElementPostprocessor,
                                    public MooseVariableInterface<Real>
    {
    public:
      static InputParameters validParams();

      ElementSumPostprocessor(const InputParameters & parameters);

      virtual void initialize() override;
      virtual void execute() override;
      virtual void threadJoin(const UserObject & y) override;
      virtual Real getValue() override;

    protected:
      virtual Real computeQpIntegral() = 0;
      virtual Real computeIntegral();

      /// Holds the solution at current quadrature points
      const VariableValue & _u;

      unsigned int _qp;

      Real _integral_value;
    };


For source file (ElementSumPostprocessor.c),

    #include "ElementSumPostprocessor.h"

    #include "libmesh/quadrature.h"

    defineLegacyParams(ElementSumPostprocessor);

    InputParameters
    ElementSumPostprocessor::validParams()
    {
      InputParameters params = ElementPostprocessor::validParams();
      params.addRequiredCoupledVar("variable", "The name of the variable that this object operates on");
      params.addClassDescription("Computes a summation of values assigned in Elements");
      return params;
    }

    ElementSumPostprocessor::ElementSumPostprocessor(const InputParameters & parameters)
      : ElementPostprocessor(parameters),
        MooseVariableInterface<Real>(this,
                                     false,
                                     "variable",
                                     Moose::VarKindType::VAR_ANY,
                                     Moose::VarFieldType::VAR_FIELD_STANDARD),
        _u(coupledValue("variable")),
        _qp(0), _integral_value(0)
    {
      addMooseVariableDependency(&mooseVariableField());
    }

    void
    ElementSumPostprocessor::initialize()
    {
      _integral_value = 0;
    }

    void
    ElementSumPostprocessor::execute()
    {
      _integral_value += computeIntegral();
    }

    Real
    ElementSumPostprocessor::getValue()
    {
      gatherSum(_integral_value);
      return _integral_value;
    }

    void
    ElementSumPostprocessor::threadJoin(const UserObject & y)
    {
      const ElementSumPostprocessor & pps = static_cast<const ElementSumPostprocessor &>(y);
      _integral_value += pps._integral_value;
    }

    Real
    ElementSumPostprocessor::computeIntegral()
    {
      if (_qp == 0){
        return _u[0];
      }
      else{
        return 0;
      }
    }


I did not compile this yet. I want to make it sure this is right or not.

Thanks.

**Reply by GiudGiud** [reply]
remove this   virtual Real computeQpIntegral() = 0;
from the header since you are not defining it

looks fine to me otherwise. See what the compiler says

**Reply by hokonkim** [reply]
When recompiling this new postprocessor, there was not an error message.

I didn't use this new post processor, but I hope this will work properly.
